### ISSUE TITLE
Dynamic columns, removing/adding columns, "this.grid is undefined"

### ISSUE THREAD

**Issue by sebdavid** [vaadin/vaadin-grid#2068]
### Description
Tested with Vaadin 14.4.1
A grid in which columns are dynamically added, regarding the selection in another grid, a JS error occurs in some case. (see demo project below)

### Expected outcome
No JS error should occur...

### Actual outcome
A JS error occurs : `Uncaught TypeError: this._grid is undefined`

This is the error occurring in the demo project I created for this issue. In my full project (which I cannot share), I got the error "Cannot read property '__getRowModel' of undefined", but the origin may be the same regarding the stacktrace : 
* in demo project :
```
Uncaught TypeError: this._grid is undefined
    __setColumnTemplateOrRenderer vaadin-grid-column.js:318
    __setColumnTemplateOrRenderer vaadin-grid-column.js:317
    _pathOrHeaderChanged vaadin-grid-column.js:465
    runMethodEffect property-effects.js:819
    runEffectsForProperty property-effects.js:157
    runEffects property-effects.js:121
    _propertiesChanged property-effects.js:1692
    _propertiesChanged client-C4B53966888EAB6CBD0D20BD246409CB.cache.js:1027
    _flushProperties properties-changed.js:354
    _flushProperties property-effects.js:1536
    _invalidateProperties property-effects.js:1506
    notifyPath property-effects.js:2049
    _createScrollerRows vaadin-grid.js:505
    _createScrollerRows vaadin-grid.js:505
```
* In my full project:
```
Uncaught TypeError: Cannot read property '__getRowModel' of undefined
    at vaadin-bundle-d7625e45a68da94758da.cache.js:210
    at Array.forEach (<anonymous>)
    at HTMLElement.__setColumnTemplateOrRenderer (vaadin-bundle-d7625e45a68da94758da.cache.js:210)
    at HTMLElement._pathOrHeaderChanged (vaadin-bundle-d7625e45a68da94758da.cache.js:210)
    at Object.N [as fn] (vaadin-bundle-d7625e45a68da94758da.cache.js:306)
    at M (vaadin-bundle-d7625e45a68da94758da.cache.js:306)
    at z (vaadin-bundle-d7625e45a68da94758da.cache.js:306)
    at HTMLElement._propertiesChanged (vaadin-bundle-d7625e45a68da94758da.cache.js:306)
    at HTMLElement.s._propertiesChanged (client-C4B53966888EAB6CBD0D20BD246409CB.cache.js:1027)
    at HTMLElement._flushProperties (vaadin-bundle-d7625e45a68da94758da.cache.js:757)
    at HTMLElement._flushProperties (vaadin-bundle-d7625e45a68da94758da.cache.js:306)
    at HTMLElement._invalidateProperties (vaadin-bundle-d7625e45a68da94758da.cache.js:306)
    at HTMLElement.notifyPath (vaadin-bundle-d7625e45a68da94758da.cache.js:306)
    at vaadin-bundle-d7625e45a68da94758da.cache.js:1310
    at Array.forEach (<anonymous>)
    at HTMLElement._createScrollerRows (vaadin-bundle-d7625e45a68da94758da.cache.js:1310)
```

### Demo project
https://github.com/sebdavid/vaadin-grid-2068-this.grid_is_undefined

The demo project is about displaying the timesheets of some projects:
* Project1 : no timesheet is associated
* Project 2: 2 timesheets are associated

### Steps to reproduce

1. Goto the Hello World view
1. Select the "Project 1" in the parent grid: nothing is displayed in the child grid (because no timesheet is associated to this project)
1. Then, select the "Project 2" in the parent grid : the JS error occurs. The dynamic columns have been added, but the data is not displayed in them.

If you refresh the page, and select first the "Porject 2", then "Project 1", then "Project 2", the error is not raised.

### Browsers Affected

- [ X] Chrome
- [ X] Firefox
- [ X] Edge chromium

**Comment by sebdavid** [comment]
Workaround suggested by Javier : recreate the grid, and replace the old one.

However the issue should be fixed.

**Comment by jcgueriaud1** [comment]
In the demo project I can reproduce the javascript error:
`__getRowModel`

The error is coming this observer `_pathOrHeaderChanged` on a disconnected column (there is one call for each columns removed).
The observer is triggered by `_createScrollerRows` in vaadin-grid,js.
Specifically this code :
```
    if (this._columnTree) {
      this._columnTree[this._columnTree.length - 1].forEach(c => c.notifyPath && c.notifyPath('_cells.*', c._cells));
    }
```
The `_columnTree` contains the old columns that has been already disconnected. (and this property is updated later)

Maybe we shouldn't notify a disconnected column?
```
if (this._columnTree) {
      this._columnTree[this._columnTree.length - 1].filter(c => c.isConnected).forEach(c => c.notifyPath && c.notifyPath('_cells.*', c._cells));
    }
```

The other problem `this_grid is not defined` is probably related (change or update a observed property on a disconnected column) as the disconnectedCallback contains `this._gridValue = undefined;` (which is the cached value of _grid)

![2068_createScrollerRows](https://user-images.githubusercontent.com/51313578/98115377-6b1b6680-1eaf-11eb-8814-501c0edef897.png)
![2068_pathOrHeaderChanged](https://user-images.githubusercontent.com/51313578/98115394-6f478400-1eaf-11eb-8c2b-5af5a701e9f8.png)

**Comment by jcgueriaud1** [comment]
I think this problem is related even if there is no steps to reproduce (on the forum):
https://vaadin.com/forum/thread/18268375/error-typeerror-cannot-read-property-__updateheaderfooterrowvisibilit

**Comment by sissbruecker** [comment]
This has been fixed as part of #1981. The fix is available from Vaadin version 14.5 and above.
